clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DeployCloudFoundryServerGroupAtomicOperationTest.java

```diff
@@ -128,6 +129,142 @@ void executeOperation() {
+        ""
+      ))
+      .setSpace(CloudFoundrySpace.builder().id("space1Id").name("space1").build())
+      .setArtifact(Artifact.builder().reference("ref1").build())
+      .setApplicationAttributes(new DeployCloudFoundryServerGroupDescription.ApplicationAttributes()
+        .setInstances(7)
+        .setMemory("1G")
+        .setDiskQuota("2048M")
+        .setHealthCheckType("http")
+        .setHealthCheckHttpEndpoint("/health")
+        .setBuildpacks(io.vavr.collection.List.of("buildpack1", "buildpack2").asJava())
+        .setServices(io.vavr.collection.List.of("service1").asJava())
+        .setEnv(HashMap.of(
+          "token", "ASDF"
+        ).toJavaMap()));
+    description.setClient(cloudFoundryClient);
+    description.setRegion("region1");
+    description.setStartApplication(true);
+    final CloudFoundryClusterProvider clusterProvider = mock(CloudFoundryClusterProvider.class);
+    final DeployCloudFoundryServerGroupAtomicOperation operation =
+      new DeployCloudFoundryServerGroupAtomicOperation(new PassThroughOperationPoller(), description, clusterProvider);
+
+    final Applications apps = cloudFoundryClient.getApplications();
+    when(clusterProvider.getClusters()).thenReturn(Collections.emptyMap());
+    when(apps.createApplication(any(), any(), any(), any()))
+      .thenReturn(CloudFoundryServerGroup.builder().id("serverGroupId").space(
+        CloudFoundrySpace.builder().id("spaceId").build()
+      ).build());
+    when(apps.getProcessState(any())).thenReturn(ProcessStats.State.CRASHED);
+    when(apps.createPackage(any()))
+      .thenAnswer((Answer<String>) invocation -> {
+        Object[] args = invocation.getArguments();
+        return args[0].toString() + "_package";
+      });
+
+    // When
+    final DeploymentResult result = operation.operate(Lists.emptyList());
+
+    // Then
+    final InOrder inOrder = Mockito.inOrder(apps, cloudFoundryClient.getServiceInstances());
+    inOrder.verify(apps).createApplication("app1-stack1-detail1-v000",
+      CloudFoundrySpace.builder().id("space1Id").name("space1").build(),
+      io.vavr.collection.List.of("buildpack1", "buildpack2").asJava(),
+      HashMap.of(
+        "token", "ASDF"
+      ).toJavaMap());
+    inOrder.verify(apps).uploadPackageBits(eq("serverGroupId_package"), any());
+    inOrder.verify(apps).createBuild("serverGroupId_package");
+    inOrder.verify(apps).scaleApplication("serverGroupId", 7, 1024, 2048);
+    inOrder.verify(apps).updateProcess("serverGroupId", null, "http", "/health");
+    inOrder.verify(cloudFoundryClient.getServiceInstances()).createServiceBindingsByName(any(), eq(Collections.singletonList("service1")));
+    inOrder.verify(apps).startApplication("serverGroupId");
+
+    assertThat(testTask.getStatus().isFailed()).isTrue();
+  }
+
+  @Test
+  void executeOperationWithNoStartFlag() {
+    // Given
+    final DeployCloudFoundryServerGroupDescription description = new DeployCloudFoundryServerGroupDescription()
```

Comment:
Are `DeployCloudFoundryServerGroupDescription` in this and prior test the same except for a property or two? Can they be extracted out of the method?

Reply:
Yes, we can probably extract that. Will fix in the morning.

Reply:
Extracted code and some other duplicated code.